File: scripts/validate-structured-data.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const MAX_SKU_LEN = 50;
const URL_LIKE_RE = /^https?:\/\//i;

const SKIP_HTML = new Set([
  'googlead697dfed14475b1.html',
  'index.original.html',
  'index.wireframe.html',
  'test_logo.html',
  'test_stats.html',
  'news/contact-us.html',
  'zh/news/contact-us.html',
]);

function walk(dir, out = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (entry.name === '.git' || entry.name === 'backend' || entry.name === 'node_modules') continue;
    const abs = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walk(abs, out);
    } else if (entry.isFile() && entry.name.endsWith('.html')) {
      const rel = path.relative(ROOT, abs).replace(/\\/g, '/');
      if (!SKIP_HTML.has(rel)) out.push(rel);
    }
  }
  return out.sort();
}

function extractJsonLdBlocks(html) {
  const blocks = [];
  const re = /<script\b[^>]*type=["']application\/ld\+json["'][^>]*>([\s\S]*?)<\/script>/gi;
  let match;
  while ((match = re.exec(html))) {
    blocks.push(match[1].trim());
  }
  return blocks;
}

function collectProducts(node, products = []) {
  if (!node || typeof node !== 'object') return products;
  if (Array.isArray(node)) {
    for (const item of node) collectProducts(item, products);
    return products;
  }
  const type = node['@type'];
  const types = Array.isArray(type) ? type : type ? [type] : [];
  if (types.includes('Product')) products.push(node);
  for (const value of Object.values(node)) {
    if (value && typeof value === 'object') collectProducts(value, products);
  }
  return products;
}

function validateProduct(product, fileRel) {
  const issues = [];
  const sku = product.sku;

  if (sku == null || sku === '') {
    issues.push({ file: fileRel, field: 'sku', issue: 'missing or empty sku' });
  } else if (Array.isArray(sku)) {
    issues.push({ file: fileRel, field: 'sku', issue: `sku is array with ${sku.length} values`, value: JSON.stringify(sku) });
    if (sku.length > 1) {
      issues.push({ file: fileRel, field: 'sku', issue: 'multiple sku values in array' });
    }
  } else {
    const s = String(sku);
    if (s.length > MAX_SKU_LEN) {
      issues.push({ file: fileRel, field: 'sku', issue: `sku longer than ${MAX_SKU_LEN} chars (${s.length})`, value: s });
    }
    if (URL_LIKE_RE.test(s) || /waikwantent\.com/i.test(s)) {
      issues.push({ file: fileRel, field: 'sku', issue: 'sku looks like a URL', value: s });
    }
    if (/\s/.test(s)) {
      issues.push({ file: fileRel, field: 'sku', issue: 'sku contains spaces', value: s });
    }
    if (/[a-z]/.test(s)) {
      issues.push({ file: fileRel, field: 'sku', issue: 'sku contains lowercase letters', value: s });
    }
    if (!/^[A-Z0-9\-_.~]+$/.test(s)) {
      issues.push({ file: fileRel, field: 'sku', issue: 'sku contains unsafe characters', value: s });
    }
  }

  if (!product.name) {
    issues.push({ file: fileRel, field: 'name', issue: 'missing Product.name' });
  }
  if (!product.image || (Array.isArray(product.image) && product.image.length === 0)) {
    issues.push({ file: fileRel, field: 'image', issue: 'missing Product.image' });
  }
  if (product.offers) {
    const offers = Array.isArray(product.offers) ? product.offers : [product.offers];
    for (const offer of offers) {
      if (!offer || typeof offer !== 'object') continue;
      if (!offer.url) {
        issues.push({ file: fileRel, field: 'offers.url', issue: 'Offer missing url' });
      }
      if (offer.price == null && !offer.priceSpecification) {
        issues.push({ file: fileRel, field: 'offers.price', issue: 'Offer missing price/priceSpecification' });
      }
    }
  }

  return issues;
}

const htmlFiles = walk(ROOT);
const allIssues = [];
let productCount = 0;

for (const rel of htmlFiles) {
  const html = fs.readFileSync(path.join(ROOT, rel), 'utf8');
  for (const block of extractJsonLdBlocks(html)) {
    let parsed;
    try {
      parsed = JSON.parse(block);
    } catch {
      allIssues.push({ file: rel, field: 'json-ld', issue: 'invalid JSON-LD block' });
      continue;
    }
    const products = collectProducts(parsed);
    for (const product of products) {
      productCount += 1;
      allIssues.push(...validateProduct(product, rel));
    }
  }
}

if (allIssues.length === 0) {
  console.log(`OK: scanned ${htmlFiles.length} HTML files, ${productCount} Product schema item(s), no issues found.`);
  process.exit(0);
}

console.error(`Found ${allIssues.length} structured-data issue(s) across ${productCount} Product item(s):\n`);
for (const item of allIssues) {
  const extra = item.value ? ` → ${item.value}` : '';
  console.error(`  [${item.file}] ${item.field}: ${item.issue}${extra}`);
}
process.exit(1);

File: scripts/validate-structured-data.py
#!/usr/bin/env python3
"""Validate Product JSON-LD sku and required merchant fields in static HTML."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
MAX_SKU_LEN = 50
SKIP = {
    "googlead697dfed14475b1.html",
    "index.original.html",
    "index.wireframe.html",
    "test_logo.html",
    "test_stats.html",
    "news/contact-us.html",
    "zh/news/contact-us.html",
}

JSONLD_RE = re.compile(
    r'<script\b[^>]*type=["\']application/ld\+json["\'][^>]*>([\s\S]*?)</script>',
    re.IGNORECASE,
)
URL_LIKE = re.compile(r"^https?://", re.I)


def walk_html() -> list[Path]:
    out: list[Path] = []
    for path in ROOT.rglob("*.html"):
        rel = path.relative_to(ROOT).as_posix()
        if rel in SKIP or ".git" in rel.split("/"):
            continue
        out.append(path)
    return sorted(out)


def collect_products(node, products=None):
    products = products or []
    if isinstance(node, list):
        for item in node:
            collect_products(item, products)
    elif isinstance(node, dict):
        t = node.get("@type")
        types = t if isinstance(t, list) else ([t] if t else [])
        if "Product" in types:
            products.append(node)
        for v in node.values():
            if isinstance(v, (dict, list)):
                collect_products(v, products)
    return products


def validate_product(product: dict, rel: str) -> list[str]:
    issues: list[str] = []
    sku = product.get("sku")

    if sku is None or sku == "":
        issues.append(f"[{rel}] sku: missing or empty")
    elif isinstance(sku, list):
        issues.append(f"[{rel}] sku: array with {len(sku)} values -> {sku!r}")
    else:
        s = str(sku)
        if len(s) > MAX_SKU_LEN:
            issues.append(f"[{rel}] sku: longer than {MAX_SKU_LEN} ({len(s)}) -> {s!r}")
        if URL_LIKE.search(s) or "waikwantent.com" in s.lower():
            issues.append(f"[{rel}] sku: looks like URL -> {s!r}")
        if re.search(r"\s", s):
            issues.append(f"[{rel}] sku: contains spaces -> {s!r}")
        if re.search(r"[a-z]", s):
            issues.append(f"[{rel}] sku: contains lowercase -> {s!r}")
        if not re.fullmatch(r"[A-Z0-9\-_.~]+", s):
            issues.append(f"[{rel}] sku: unsafe characters -> {s!r}")

    if not product.get("name"):
        issues.append(f"[{rel}] name: missing Product.name")
    image = product.get("image")
    if not image or (isinstance(image, list) and len(image) == 0):
        issues.append(f"[{rel}] image: missing Product.image")

    offers = product.get("offers")
    if offers:
        offer_list = offers if isinstance(offers, list) else [offers]
        for offer in offer_list:
            if not isinstance(offer, dict):
                continue
            if not offer.get("url"):
                issues.append(f"[{rel}] offers.url: missing")
            if offer.get("price") is None and not offer.get("priceSpecification"):
                issues.append(f"[{rel}] offers.price: missing price/priceSpecification")

    return issues


def main() -> int:
    files = walk_html()
    all_issues: list[str] = []
    product_count = 0

    for path in files:
        rel = path.relative_to(ROOT).as_posix()
        html = path.read_text(encoding="utf-8", errors="replace")
        for block in JSONLD_RE.findall(html):
            try:
                parsed = json.loads(block.strip())
            except json.JSONDecodeError:
                all_issues.append(f"[{rel}] json-ld: invalid JSON block")
                continue
            for product in collect_products(parsed):
                product_count += 1
                all_issues.extend(validate_product(product, rel))

    if not all_issues:
        print(
            f"OK: scanned {len(files)} HTML files, {product_count} Product schema item(s), no issues found."
        )
        return 0

    print(f"Found {len(all_issues)} issue(s) across {product_count} Product item(s):\n")
    for issue in all_issues:
        print(f"  {issue}")
    return 1


if __name__ == "__main__":
    sys.exit(main())
